sdk/nodejs/serverless/function.ts

```diff
@@ -0,0 +1,184 @@
+// Copyright 2016-2018, Pulumi Corporation.
```

Comment:
Although a little roundabout - the way we add these "overlays" currently is via a mechanism in the `resources.go` that copies them into the generated folder instead of authoring directly in the generated folder.  See https://github.com/pulumi/pulumi-aws/blob/master/resources.go#L1831 for the equivalent in the AWS provider.

Reply:
Not sure I understand what you mean here... What are "overlays"?

Reply:
@mikhailshilkov So, a good example of this is:

https://github.com/pulumi/pulumi-aws/blob/master/overlays/nodejs/ec2/instanceType.ts

The idea here is that *after* generating all the auto-generated code for our Resource types, we can add those 'overlay' files to the resultant folders to expose this extra API functionality through that package.   So, if you look up into https://github.com/pulumi/pulumi-aws/tree/master/sdk/nodejs/ec2, you'll see only that single file.  But if you then look at:

https://github.com/pulumi/pulumi-aws/tree/master/sdk/nodejs/ec2

You'll see the entire generated EC2 API *along with* that 'instanceType.ts' file.

--

So, the 'overlays' mechanism allows us to auto generate a large API, while also allowing us to 'inject' files into the final result.

Reply:
I still don't get it , sorry. The last 2 links seem to be the same, and `instanceType.ts` seems the same in `overlay` and non-overlay folders...
What should I do?

Reply:
It's a tad convoluted, so my apologies for the confusion.

Basically, in the AWS repo you have two files for the serverless equivalent:

* [`overlays/nodejs/serverless/function.ts`](https://github.com/pulumi/pulumi-aws/blob/master/overlays/nodejs/serverless/function.ts) - the file you write by hand
* [`sdk/nodejs/serverless/function.ts`](https://github.com/pulumi/pulumi-aws/blob/master/sdk/nodejs/serverless/function.ts) - the file that is generated when you `make`

Basically, the entire `sdk/*` directory is generated when you run `make`, mostly from Terraform metadata, however it also mixes in directives from the [`resources.go`](https://github.com/pulumi/pulumi-aws/blob/master/resources.go) file. In this specific case, there is [a directive in that file](https://github.com/pulumi/pulumi-aws/blob/4c982c2f275c6b39383acac761f44449c099b461/resources.go#L1831) which is responsible for merging that file in `overlays/...` into the resulting built `sdk/...` directory. As a result, everything underneath `sdk/*` is generated.

I do get the benefit of being consistent here, although honestly, @lukehoban, @CyrusNajmabadi, we may want to revisit this now that we commit the generated sdk directories...partly, we did it this way originally because we didn't commit them, although there are still possibly reasons (e.g., it's clearer what is generated versus not).

At the end of the day, I would *love* to land this change, so I'm happy to help refactor it to conform to this alternative model, provided all is working and otherwise well.

**Thank you** for doing this! 🎉

Reply:
Ah, okay, I'll move it to `overlays` folder.

Reply:
Wait, there's no `overlays` folder in GCP. I'm confused again about what needs to happen.